Fix issues with pip

twangboy opened the pull request:
### What does this PR do?
The pip module has been basically broken since pip 8.1.2 due to how pip has changed the output of some functions. It has also been problematic in Windows since `c:\salt\bin` and `c:\salt\bin\Scripts` are not a part of the search path. Therefore you always had to pass the full path to the pip binary.

This PR does many things:
1. Defaults to the pip for the running python if `bin_env` is not passed. Uses `python -m pip`
2. Defaults to the cwd for the running python if `cwd` is not passed on Windows
3. Uses a list to build commands throughout
4. Fixes a problem with parsing lines that have `#egg=` in them in the `pip.list` function
5. Fixes issues where `pip --version` was being called many times
6. Fixes an issue in `pip.list_upgrades` where the output format has changed since version pip 8.0.0.
7. Also adds support for the `--format json` option available in pip 9.0.0 so we don't have to parse stdout
8. Skips `salt` in `pip.upgrade` function on Windows as it will fail to update itself

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/47024
https://github.com/saltstack/salt/issues/47146

### Tests written?
No

### Commits signed with GPG?
Yes

twangboy committed:
Fix merge conflict

twangboy commented:
I'll probably have to fix tests... ran out of time today

rallytime reviewed (changes requested):
Yeah the tests need to be addressed here and there's some failing lint: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/21147/violations/file/salt/modules/pip.py/

:D

twangboy committed:
Fix lint error

gtmanfred commented on salt/modules/pip.py:
```diff
@@ -101,6 +102,7 @@
 }
 
 VALID_PROTOS = ['http', 'https', 'ftp', 'file']
+PIP_VERSION = None
```
This seems like it could blow up in our faces if someone upgrades pip manually with `pip install -U pip`.

This I think would be a good use case for `__context__['PIP_VERSION']` it will have to be run each time the module is loaded, but it would also be able to be cleared out by a `saltutil.refresh_modules`

twangboy committed:
Fix some tests``

rallytime commented:
@twangboy Lint: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/21310/violations/file/tests/unit/modules/test_pip.py/

vernondcole commented on salt/modules/pip.py:
```diff
@@ -120,37 +122,58 @@ def _get_pip_bin(bin_env):
+
+    # try to get python bin from virtualenv, bin_env
+    python_bin = 'python'
+    if salt.utils.is_windows():
```
How about:
`python_bin = 'python.exe' if salt.utils.is_windows() else 'python`?
  It saves two source lines, eliminates an unneeded assignment (on Windows)
and is easier to read and understand.

twangboy committed:
Change from global variable to __context__``

terminalmage committed:
DRY and other fixes in pip module

This makes the following adjustments:

- DRY fix for repeated search paths
- Checks that python bin found in virtualenv is executable
- Removes string-escape encoding, it doesn't have any use if we're not
  dealing with quotes/special characters
- Removes redundant islink check, since a symlink will return True when
  it points to a file and not a directory, and we know thanks to the
  prior if block that the path is not a directory so it can only be a
  file.
- `pip` is almost never referred to in upper case, log messages changed
  to reflect this.
- Make context caching of pip version specific to bin_env

terminalmage reviewed (commented):
I talked with @twangboy and submitted some changes in https://github.com/twangboy/salt/pull/8 to fix some failing tests, and clean up `_get_pip_bin()` a bit.

terminalmage commented:
@twangboy can you rebase?

twangboy commented:
@terminalmage Rebased...

terminalmage committed:
Lint fix

terminalmage committed:
Update pip unit tests to reflect changes

terminalmage commented:
Opened https://github.com/twangboy/salt/pull/9 against @twangboy's fork to fix lint and test failures

twangboy committed:
Merge pull request #9 from terminalmage/twangboy/fix_47024

Lint and test fixes

twangboy committed:
Fix integration.modules.test_pip

Removes references to `no_chown`. It hasn't been in use since 2015.8
Removes usage of user in pip calls. The user is defined as the current
user which is the default anyway and it's causing tests not to run in
Windows because Windows requires a password
Removes the pwd import since it is no longer needed

terminalmage commented:
@twangboy I think that the context caching stuff may not be feasible for the test suite, still another test failure related to it: https://jenkins.saltstack.com/job/PR/job/salt-pr-rs-cent7-n/18522/testReport/junit/integration.modules.test_pip/PipModuleTest/test_issue_2087_missing_pip/

We may need to revisit this. Also, this needs another rebase.

twangboy committed:
Fix integration.modules.test_pip

Pass password as kwarg for runas on Windows
Fix issues with pip 10 in unit.states.test_pip

twangboy committed:
Merge branch '2017.7' into fix_47024

twangboy committed:
Fix typo

twangboy committed:
Add @with_tempdir helper

Fix two state tests
Remove redundent tests

rallytime commented:
@twangboy This is failing some pip tests in PY2: https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/22484/

twangboy commented:
@rallytime I think it's failing because the test suite is using virtualenv 1.10 from 2013 which is incredibly old. The pip environment inside doesn't the `--no-cache-dir` option. I submitted a PR: https://github.com/saltstack/salt-jenkins/pull/957

rallytime commented:
re-run py2

gtmanfred commented:
This change to virtualenv is now causing this test to fail https://jenkinsci.saltstack.com/job/2017.7/job/salt-fedora-26-py2/125/testReport/junit/integration.states.test_virtualenv/VirtualenvTest/test_issue_1959_virtualenv_runas/

rallytime commented:
@twangboy Can you take a look?

rterbush commented:
Not clear to me if this PR has made it into 2019.2.0 but I am seeing the following behavior there:

Following yaml:
```yaml
pywinauto:
  pip.installed:
    - cwd: 'C:\salt\bin\scripts'
    - bin_env: 'C:\salt\bin\scripts\pip.exe'
    - upgrade: True
```
Errors with, and seems to incorrectly report success:
```
wnode1:
----------
          ID: install_pywinauto
    Function: pip.installed
        Name: pywinauto
      Result: None
     Comment: Error installing 'pywinauto': Could not find a `pip` binary
     Started: 11:54:02.820638
    Duration: 31.217 ms
     Changes:

Summary for wnode1
------------
Succeeded: 1 (unchanged=1)
Failed:    0
------------
Total states run:     1
Total run time:  31.217 ms
```
Removal of both cwd: and bin_env: allows this to succeed.
```yaml
pywinauto:
  pip.installed:
    - upgrade: True
```
Success:
```
wnode1:
----------
          ID: install_pywinauto
    Function: pip.installed
        Name: pywinauto
      Result: True
     Comment: Python package pywinauto was already installed
              All specified packages are already installed and up-to-date
     Started: 11:54:27.117672
    Duration: 2531.214 ms
     Changes:

Summary for wnode1
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1
Total run time:   2.531 s
```

Running Python3 minion on Windows, version 2019.2.0
Running Python2 master on Centos 7:
```
Salt Version:
           Salt: 2019.2.0

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.5 (default, Apr  9 2019, 14:30:50)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.6.1810 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-957.10.1.el7.x86_64
         system: Linux
        version: CentOS Linux 7.6.1810 Core
```
